File: mac/Core/Network/TlsSession.swift
import Foundation
import Network

/// TLS 1.3 control connection between two peers.
/// TODO(week 1): implement using NWConnection with NWProtocolTLS.Options + pinned trust evaluation
/// per docs/security.md. PairingService will provide the pinned SPKI hash for the verify-block.
public final class TlsSession {
    public enum TlsSessionError: Error, CustomStringConvertible {
        case notImplemented(String)

        public var description: String {
            switch self {
            case .notImplemented(let operation):
                return "TlsSession.\(operation) is not implemented in the current dev build"
            }
        }
    }

    public enum State {
        case idle
        case connecting
        case handshaking
        case ready
        case failed(Error)
        case closed
    }

    public private(set) var state: State = .idle

    public init() {}

    public func connect(host: String, port: UInt16) async throws {
        // TODO: NWConnection + NWProtocolTLS.Options, pinned-key verify, exporter extraction.
        throw TlsSessionError.notImplemented("connect")
    }

    public func send(_ data: Data) async throws {
        throw TlsSessionError.notImplemented("send")
    }

    public func close() {
        state = .closed
    }
}

File: windows/Core/Network/TlsSession.cs
namespace BarelyReal.Core.Network;

/// TLS 1.3 control connection between two peers.
/// TODO(week 1): implement using TcpClient + SslStream with custom RemoteCertificateValidationCallback
/// that enforces the pinned SPKI hash from <see cref="PairingService"/>.
public sealed class TlsSession
{
    public sealed class TlsSessionException : Exception
    {
        public TlsSessionException(string operation)
            : base($"TlsSession.{operation} is not implemented in the current dev build")
        {
        }
    }

    public enum SessionState
    {
        Idle,
        Connecting,
        Handshaking,
        Ready,
        Failed,
        Closed
    }

    public SessionState State { get; private set; } = SessionState.Idle;

    public Task ConnectAsync(string host, ushort port, CancellationToken ct = default)
    {
        // TODO: TcpClient + SslStream.AuthenticateAsClientAsync with SslClientAuthenticationOptions
        //       { EnabledSslProtocols = SslProtocols.Tls13, RemoteCertificateValidationCallback = pin-check }
        throw new TlsSessionException("ConnectAsync");
    }

    public Task SendAsync(ReadOnlyMemory<byte> data, CancellationToken ct = default)
    {
        throw new TlsSessionException("SendAsync");
    }

    public void Close()
    {
        State = SessionState.Closed;
    }
}
